storage/range.go

```diff
@@ -1451,60 +1497,60 @@ func (r *Range) InternalTruncateLog(batch engine.Engine, ms *engine.MVCCStats, a
 	reply.SetGoError(err)
 }
 
-// InternalLeaderLease evaluates and responds to a request to grant a leader lease.
-func (r *Range) InternalLeaderLease(args *proto.InternalLeaderLeaseRequest, reply *proto.InternalLeaderLeaseResponse) {
-	// TODO(tschottdorf) stub for now to get tests working.
-	r.setLease(&args.Lease)
-	// r.grantLeaderLease(args.Lease)
-}
+// InternalLeaderLease sets the leader lease for this range. After a
+// lease has been set, calls to HasLeaderLease() will return true if
+// this replica is the lease holder and the lease has not yet
+// expired. If this range replica is already the lease holder, the
+// expiration will be extended as indicated. Otherwise, all duties
+// required of the range leader are commenced, including gossiping
+// cluster and config info where appropriate and clearing the command
+// queue and timestamp cache.
+func (r *Range) InternalLeaderLease(batch engine.Engine, ms *engine.MVCCStats, args *proto.InternalLeaderLeaseRequest, reply *proto.InternalLeaderLeaseResponse) {
+	r.Lock()
+	defer r.Unlock()
 
-// requestLeaderLease sends a request to obtain or extend a leader lease for
-// this replica. Being a first mover, it registers itself as a task with the
-// stopper.
-func (r *Range) requestLeaderLease(term uint64) {
-	if !r.stopper.StartTask() {
+	// Set the new leader lease.
+	oldLease := r.getLease()
+	if args.Lease.Term < oldLease.Term {
+		reply.SetGoError(util.Errorf("lease term %d older than last granted lease %d", args.Lease.Term, oldLease.Term))
 		return
 	}
-	defer r.stopper.FinishTask()
-	// Prepare a Raft command to get a leader lease for the replica
-	// of that group that lives in our store.
-	wallTime := r.rm.Clock().PhysicalNow()
-	// TODO: get this from configuration, either as a config flag
-	// or, later, dynamically adjusted.
-	duration := int64(defaultLeaderLeaseDuration)
-	idKey := makeCmdIDKey(proto.ClientCmdID{
-		WallTime: wallTime,
-		Random:   rand.Int63(),
-	})
-	cmd := proto.InternalRaftCommand{
-		RaftID: r.Desc().RaftID,
-	}
-	args := &proto.InternalLeaderLeaseRequest{
-		RequestHeader: proto.RequestHeader{Key: r.Desc().StartKey},
-		Lease: proto.Lease{
-			Expiration: wallTime + duration,
-			Duration:   duration,
-			Term:       term,
-			RaftNodeID: uint64(r.rm.RaftNodeID()),
-		},
+	// Store the lease to disk & in-memory.
+	if err := engine.MVCCPutProto(batch, nil, engine.RaftLeaderLeaseKey(r.Desc().RaftID), proto.ZeroTimestamp, nil, &args.Lease); err != nil {
```

Comment:
This is too late to persist the lease. Like the HardState, the leader lease must be persisted before we send back the MsgAppResp acknowledging it (which happens in raft). We need to add something to the multiraft storage interface so that multiraft can call back into the storage layer when it sets leaseGrantedUntil.